ledger-core/v2/reference/marshal_local.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2020 Insolar Network Ltd.
+// All rights reserved.
+// This material is licensed under the Insolar License version 1.0,
+// available at https://github.com/insolar/assured-ledger/blob/master/LICENSE.md.
+
+package reference
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+
+	"github.com/insolar/assured-ledger/ledger-core/v2/vanilla/throw"
+)
+
+func ProtoSizeLocal(h LocalHolder) int {
+	if h == nil || h.IsEmpty() {
+		return 0
+	}
+	v := h.GetLocal()
+	switch pn := v.GetPulseNumber(); {
+	case pn.IsUnknown():
+		return 0
+	case pn.IsTimePulse():
+		return LocalBinarySize
+	}
+	return LocalBinaryPulseAndScopeSize + v.hashLen()
+}
+
+func MarshalLocal(h LocalHolder) ([]byte, error) {
+	if h == nil || h.IsEmpty() {
+		return nil, nil
+	}
+	b := make([]byte, LocalBinarySize)
+	n, err := MarshalLocalTo(h, b)
+	if err != nil {
+		return nil, err
+	}
+	return b[:n], nil
+}
+
+func MarshalLocalTo(h LocalHolder, data []byte) (int, error) {
+	if h == nil || h.IsEmpty() {
+		return 0, nil
+	}
+	v := h.GetLocal()
+	switch pn := v.GetPulseNumber(); {
+	case pn.IsUnknown():
+		return 0, nil
+	case pn.IsTimePulse():
+		if len(data) >= LocalBinarySize {
+			return WriteWholeLocalTo(v, data), nil
+		}
+	case len(data) >= LocalBinaryPulseAndScopeSize:
+		copy(data, v.pulseAndScopeAsBytes())
+		i := v.hashLen()
+		if copy(data[LocalBinaryPulseAndScopeSize:], v.hash[:i]) == i {
+			return LocalBinaryPulseAndScopeSize + i, nil
+		}
+	}
+	return 0, throw.WithStackTop(io.ErrShortBuffer)
+}
+
+func UnmarshalLocalJSON(data []byte) (v Local, err error) {
+	var repr interface{}
+
+	err = json.Unmarshal(data, &repr)
+	if err != nil {
+		err = throw.W(err, "failed to unmarshal reference.Local")
+		return
+	}
+
+	switch realRepr := repr.(type) {
+	case string:
+		var decoded Global
+		decoded, err = DefaultDecoder().Decode(realRepr)
+		if err != nil {
+			err = throw.W(err, "failed to unmarshal reference.Local")
+			return
+		}
+		// TODO must check and fail if it was a full reference
+		v = decoded.GetLocal()
+	case nil:
+		//
+	default:
+		err = throw.W(err, "unexpected type %T when unmarshal reference.Local", repr)
+	}
+	return
+}
+
+func MarshalLocalJSON(v LocalHolder) (b []byte, err error) {
```

Comment:
The method actually marshall ref to string form

Reply:
well, it mimics MarshalJSON signature for simpler use, hence there are []byte here